text/0000-or-patterns.md

```diff
@@ -0,0 +1,666 @@
+```
+
+And for `while let` expressions:
+
+```rust
+expr_while_let : maybe_label WHILE LET top_pat '=' expr_nostruct block ;
+```
+
+For `for` loop expressions we now have:
+
+```rust
+expr_for : maybe_label FOR top_pat IN expr_nostruct block ;
+```
+
+For `match` expressions we now have:
+
+```rust
+expr_match : MATCH expr_nostruct '{' match_clause* nonblock_match_clause? '}' ;
+match_clause : nonblock_match_clause ',' | block_match_clause ','? ;
+nonblock_match_clause : match_arm (nonblock_expr | block_expr_dot) ;
+block_match_clause : match_arm (block | block_expr) ;
+
+match_arm : maybe_outer_attrs top_pat (IF expr_nostruct)? FAT_ARROW ;
+```
+
+In other words, in all of the contexts where a pattern is currently accepted,
+the compiler will now accept pattern alternations of form `p | q` where
+`p` and `q` are arbitrary patterns.
+
+## Static semantics
+
+1. Given a pattern `p | q` at some depth for some arbitrary patterns `p` and `q`,
+   the pattern is considered ill-formed if:
+
+   + the type inferred for `p` does not unify with the type inferred for `q`, or
+   + the same set of bindings are not introduced in `p` and `q`, or
+   + the type of any two bindings with the same name in `p` and `q` do not unify.
+
+2. When type checking an expression `match e_s { a_1 => e_1, ... a_n => e_n }`,
+   for each match arm `a_i` which contains a pattern of form `p_i | q_i`,
+   the pattern `p_i | q_i` is considered ill formed if,
+   at the depth `d` where it exists the fragment of `e_s` at depth `d`,
+   the type of the expression fragment does not unify with `p_i | q_i`.
+
+3. With respect to exhaustiveness checking, a pattern `p | q` is
+   considered to cover `p` as well as `q`. For some constructor `c(x, ..)`
+   the distributive law applies such that `c(p | q, ..rest)` covers the same
+   set of value as `c(p, ..rest) | c(q, ..rest)` does. This can be applied
+   recursively until there are no more nested patterns of form `p | q` other
+   than those that exist at the top level.
+
+   Note that by *"constructor"* we do not refer to tuple struct patterns,
+   but rather we refer to a pattern for any product type.
+   This includes enum variants, tuple structs, structs with named fields,
+   arrays, tuples, and slices.
+
+## Dynamic semantics
+
+1. The dynamic semantics of pattern matching a scrutinee expression `e_s`
+   against a pattern `c(p | q, ..rest)` at depth `d` where `c` is some constructor,
+   `p` and `q` are arbitrary patterns, and `rest` is optionally any remaining
+   potential factors in `c`, is defined as being the same as that of
+   `c(p, ..rest) | c(q, ..rest)`.
+
+## Implementation notes
+
+With respect to both static and dynamic semantics,
+it is always valid to first desugar a pattern `c(p | q)`
+in CNF to its equivalent form in DNF, i.e. `c(p) | c(q)`.
+
+However, implementing `c(p | q)` in terms of a pure desugaring to `c(p) | c(q)`
+may not be optimal as the desugaring can result in exponential blow-up of patterns.
```

Comment:
Please give a concrete example here. (Also, s/exponential/multiplicative/.)

For an example, `(0|1, 0|1, 0|1, 0|1, ...)` seems like the simplest one to understand.